### PR TITLE
Add post: Evaluation-Driven Agent Readiness in Copilot Studio

### PR DESCRIPTION
**Pull request by KarimaKT**
## Summary
- Introduces an evaluation-driven approach to agent readiness
- Organizes readiness around coverage, explanation, and direction
- Shows how to scope test sets, select graders, interpret failures, and iterate
- Includes a hands-on Copilot Studio walkthrough with complete screenshots

## Checklist
- [x] Review feedback addressed
- [x] All images include alt text and captions
- [x] Contextual internal links added
- [x] Tags selected for relevant Further Reading results
- [x] Post marked for publication
- [ ] Latest PR preview build passes

### CODE REVIEW

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+The fix is to make evaluation part of your design work as the maker, not a gate someone else bolts on at the end. You run it as a loop: generate targeted tests, run them, read what they tell you, improve the agent, and run again. Every loop gives you evidence instead of opinions. You stop debating whether the agent is *good* and start showing whether this version beats the last one, and exactly where.
+
+> The single biggest lever on test quality is how well you've shaped the agent itself. If you haven't already, [Influencing Agent Planning with Contextual Instructions]({% post_url 2025-11-11-influence-orchestration-knowledge %}) is a good companion read, because a sharply scoped agent produces sharper evaluations.
+{: .prompt-info }
+
+## A better question than "is it good?"
+
+"Is my agent good?" feels like the right question, but it has no answer. A generative agent reasons over tools, knowledge, instructions, and an open-ended range of phrasings, so any single overall score hides more than it shows. Tell a stakeholder the agent scored 78% and you'll get a fair-sounding question back that nobody can answer: is that good?
```
Comment (adilei) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: The problematic question is "is my agent good *enough*"

The *enough* is the problematic part

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+And a generative agent is statistical, so "perfect" was never the bar. "Is it good?" really means "is it good enough to ship?", and that splits into two questions you *can* answer: did I build the right things, the right tools, knowledge, and instructions, and have I tested them against the cases that matter, including the edge cases and the ones that should fail?
```
Comment (adilei) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: We should flesh out the problem a little bit more here.

The reason why focusing on a threshold (e.g. 80% success rate) is problematic is because thresholds do not have fixed meaning. Thresholds are meaningful only in relation to your test set size, coverage %, and how the different test cases are weighted. The same 80% can be noise on 20 rows or signal on 2,000, and most customers do not have the luxury of having data science teams defining their eval strategy for each and every agent.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+A better question is narrower, and really two in one:
+
+> Does this agent cover what I expect, with quality I can explain? And is this version better than the last, in exactly which scenarios?
```
Comment (adilei) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: This section is the heart of the post, but it's underdeveloped and the "narrower / two in one" framing undersells it. The point isn't that the question gets *narrower*. It's that the single threshold question should be replaced by a few distinct questions, each targeting something a threshold can't answer on its own:

1. **Does this version improve on the most recent one, and in exactly which scenarios?** 
2. **Does my test set have enough coverage to trust the result?** (size, representativeness, and weighting are what give *any* number its meaning)
3. **When a case fails, *why* did it fail?** Is it a real bug, an acceptable miss, or correct behavior like a guardrail refusal? Interpret failures by cause, not just as a flat pass/fail.

Let's flesh this out and make it the spine of the article: state these three questions explicitly here, then have the rest of the post visibly answer them. Generation + coverage answers #2, 'Compare with' answers #1, and reading grader explanations answers #3. Those ideas already exist later in the post, but they're scattered as techniques. Pulling them up into this section as the organizing frame would make everything downstream feel like it's answering a question we posed, instead of a list of features.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+```mermaid
+flowchart LR
+    A[Define a narrow agent] --> B[Generate targeted tests]
+    B --> C[Run the evaluation]
+    C --> D[Read the failures]
+    D --> E[Improve the design]
+    E --> F[Rerun the same tests]
+    F -->|new failure pattern| B
+    F -->|better, with known gaps| G[Explainable readiness]
+```
```
Comment (adilei) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: Minor flow note: this diagram interrupts the momentum a little. We've just declared what should replace the threshold, so the natural next move is to dig into those alternatives in more detail rather than pause on a high-level loop. Might read more smoothly if we go straight into them, and bring back a visual later only if it earns its place.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+_The evaluation-driven design loop. You leave it when coverage is understood, agreed, and explainable, not when a number is high._
+
+## Set the stage: a narrow shipping agent
```
Comment (adilei) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: Worth separating two problems that are tangled together here, because the post actually addresses both:

1. **Agreement** — do makers and business stakeholders ever agree what "good"/"ready" means? (a process gap)
2. **Method** — even when they try, is "threshold + test cases" a defensible way to answer it? (a measurement gap)

We can touch on both, but better not at length — just be explicit that there are two. Right now they're blended, so the reader doesn't notice.

If we're using this example to set the scene, let's be explicit about what it's actually doing: **before we can talk about the evaluation method at all, there has to be agreement on what the agent does and doesn't do.** That agreement is the foundation, and being clear about it here pays off later — coverage, guardrails, and reading failures by cause all build on it. As written, the example reads as a standalone "narrow your agent" tip rather than step one of that foundation.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+
+Two tools back the role: `GetShipmentStatus` (by tracking number or shipment ID) and `LookupOrderDetails` (by order number). A small SharePoint-backed knowledge set covers the status glossary and the escalation policy, so when a case needs a human the agent points the user to the right path from policy instead of acting on the shipment itself. (For the mechanics of wiring up that knowledge, see [Connecting to Custom Knowledge Sources in Copilot Studio]({% post_url 2025-09-24-copilot-studio-custom-knowledge-source %}).)
+
+![Tools list showing the two intentional tools](/assets/posts/evaluation-driven-agent-readiness/02-tools-list.png){: .shadow }
+_Keep the tool list intentionally small. Every tool you add is another path your tests have to cover._
+
+#### Why narrow beats broad
+
+A narrow role means out-of-box test generation has something concrete to reason about, your buckets make sense to both makers and stakeholders, and a failure usually points at a specific design decision rather than a vague "it's not smart enough."
+
```
Comment (adilei) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:80]: We're starting to introduce new categories here that don't actually help answer the three questions we raised, so the section reads as a fresh taxonomy rather than a step in the argument. Better to stay inside the scope we set and cash out our initial commitments:

- **Foundational core** and **Guardrails** both earn their place: they're how we ensure coverage of the **must-haves** — the must-*do* and must-*not-do* that fall straight out of the scope agreement.
- **Robustness** (edge cases, typos) is *also* a coverage move: it broadens coverage across messy, real-world phrasing. That's fine, but let's say so explicitly — it's another way to increase coverage, not a different kind of thing.
- **Architecture** is really grading methodology, not coverage. "Did it use the right tool / cite the right source?" is something you *grade* (the Tool use grader), not a slice of the input space. Two issues: the word "Architecture" doesn't communicate "test which tools get used," and even if it did, it's the wrong point to raise while we're talking about coverage. I'd move it to the graders section.

Net: keep core / guardrails / robustness as coverage (and name robustness as such), and relocate the tool-use idea to grading.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+add lower-priority coverage.
+```
+
+![Generating a test set from the agent definition](/assets/posts/evaluation-driven-agent-readiness/03-generate-test-set.png){: .shadow }
+_The Evaluate tab generates a generic starter set and, just as useful, the row format you'll reuse when you generate bucketed rows with your own LLM._
+
+> The handful of requests you wrote during use-case scoping are a great starting point. Use them to seed the larger generated sets, so every evaluation run feeds rapid, evidence-based improvement instead of guesswork.
+{: .prompt-tip }
+
+#### Combine generated rows with must-work rows
```
Comment (adilei) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: This section speaks about two separate things, and it might read more clearly if we tease them apart: **what** should be tested (must-haves vs. broad coverage) and **how** we generate the cases (hand-curated vs. LLM-generated). Right now they're a bit interleaved, so the mapping between them is easy to miss.

One way to make it flow:

1. **What you need:** must-haves *and* broad coverage. (Ties back to the Q2 coverage point — must-haves carry high weight, broad coverage fills the long tail.)
2. **How to produce each in the product:** hand-write the must-haves manually, and lean on generated questions to scale broad coverage.

This also lines up nicely with the data-science doc — *"Manual for precision, generated for coverage."* So we could land it as one clear line: **generate must-haves manually, use generated questions for coverage.**

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+
+And the **Robustness** set, with the same facts wrapped in messier language:
+
+| Question | Expected response |
+| --- | --- |
+| where's tracking C31VU3ZPDY at right now? | In Transit |
+| status of shippment New York to Tokyo | In Transit |
+| Give me the status and days in transit for the Chicago to London shipment. | In Transit; 4 days in transit |
+| Is the LA to London package on time or running late? | It is delayed |
+
```
Comment (adilei) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:155]: Two things here:

1. Up to now we've used **weight** to mean *importance* (how much a failure matters). This section reuses the same word for *density of test cases* (how many rows a bucket gets). Same term, different meaning — that's confusing. Worth picking one.
2. Is it actually true that must-haves get *more* rows? If anything I'd expect fewer: a must-have that fails should arguably fail the whole test (that's the point of a must-have), so it doesn't need volume to *gate*. The nuance is you may still want a handful of phrasings to prove it holds under real-world wording. Either way, row-count feels like the wrong axis here — must-haves are about gating weight, not density.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+#### Weight rows by what's at stake
+
+Not every bucket deserves the same number of rows. Give the broadest variation to your most frequent scenarios, deeper validation to your highest-value outcomes, and the strictest checks to your highest-risk paths. Coverage is a budget, so spend it where a failure would hurt most.
+
+## Graders, done properly
+
+A test row is only as good as the grader judging it. Copilot Studio's graders fall into two families, and the difference drives how you author your test content.
+
+| Family | Graders | Needs an expected output? |
+| --- | --- | --- |
```
Comment (adilei) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:177]: "Deterministic (**path and fact checks**)" — "fact checks" is misleading. These graders (Exact match, Keyword match, Tool use) match the response against an expected value, keyword, or tool *you* supply. They confirm the answer conforms to a known key; they don't verify whether a claim is actually true. A reasonable reader will read "fact checking" as "the grader determines if the fact is correct," which isn't what's happening — the "fact" only counts as right because you already encoded the expected answer. Worth relabeling, e.g. "exact-value / path matching," to avoid implying truth-verification.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+| **Deterministic** (path and fact checks) | Exact match, Keyword match, Tool use | Usually yes (the expected value, keyword, or tool) |
+| **AI** (quality and meaning checks) | General quality, Compare meaning, Text similarity, Custom | Mixed: Compare and Text similarity need one; General quality does not |
+
```
Comment (adilei) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:181]: Is this saying grade must-haves deterministically? If so, there's a tradeoff worth naming: **Exact match fails a *correct* answer that's just phrased differently** ("In Transit" vs. "The shipment is currently in transit") — a false negative that punishes good behavior.

The data-science doc actually steers the other way for known-good answers: it recommends **Compare meaning** (tunable pass score), because "an answer can be phrased in different correct ways but the meaning still needs to come through." It never recommends Exact match. So I'd reserve Exact match for the truly fixed cases you already list (compliance line, ID, identity), and lean on Compare meaning for must-haves where the answer matters but wording can vary.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+So which family do you pick? It depends on the answer. Some answers are too important to grade on quality or meaning alone. For the few that must be identical every time, a compliance line, a fixed escalation contact, the agent's own identity, make the response deterministic and check it with **Exact match** rather than asking an AI grader whether it came close enough.
+
+Match the grader to the kind of answer, too. Transactional answers, a status, a date, a number, check cleanly with **Exact match** or **Keyword match**. Analytical or generated answers, a summary or an explanation, need meaning-based checks like **Compare meaning** or **General quality**, and sometimes a structural check on length or format. The shipping agent is mostly transactional, which is why its core sets lean deterministic, but the moment an agent starts summarizing or reasoning, the grader mix has to shift with it.
+
+Analytical answers are where a single grader is most likely to mislead you. A knowledge-grounded summary can read perfectly and still be wrong, or right for the wrong reason, so check both the path the agent took and the answer it produced:
+
```
Comment (adilei) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:187]: Quick clar: are we suggesting stacking graders on *every* test case, or just the must-haves? It's ambiguous — line 190 says "stack on any test… per use case and per bucket" (sounds like everywhere), but the info-box just below says "one path + one answer grader per set, add more only when needed" (sounds selective). Those pull in different directions.

I'd make it explicit that stacking is **selective**: it pays off where both the *path* and the *answer* can fail independently — knowledge-grounded / tool-using answers, i.e. your high-value must-haves. For a simple guardrail refusal or a plain transactional lookup, one well-chosen grader is enough. That keeps stacking concentrated where it earns its cost instead of implying you double-grade everything.

**Review by adilei — COMMENTED** [review]
my comments inline. Thanks Karima!

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 1 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: Addressed: the "One score can't tell you it's ready" section now pivots off exactly this "enough" problem.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 2 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: Fleshed out in "One score can't tell you it's ready": thresholds as an older-world stand-in, and a pass rate over a small, curated set isn't statistically meaningful and was never meant to be.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 3 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: Done — the three questions are now Coverage / Explanation / Direction, stated up front as the frame the rest of the post answers, and restated as statements in "Explainable readiness, not a magic number."

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 4 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: Removed the high-level loop diagram; the section now goes straight into the alternatives.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 5 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: Split explicitly in "Two problems hide inside \"is it ready?\"" — a design/scope problem and an agreement problem, with the scoping example framed as step one of that foundation.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 6 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:80]: Reworked — coverage is now Must-do (robustness folded in and named as coverage) / Must-not-do guardrails / Nice-to-have, and the tool-use idea moved to "Match the grader to the job."

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 7 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: Teased apart in the "What to test versus how to produce it" subsection: hand-write the must-dos, generate to scale coverage.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 8 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:155]: Fixed — "weight" now only means importance, and the row-count/density framing is gone; must-dos gate rather than getting more rows.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 9 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:177]: Relabeled — the "fact checks" wording is gone; deterministic graders are described as matching the response against an expected value, keyword, or tool you supply.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 10 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:181]: Named the tradeoff — "Prefer Compare meaning over Exact match" with the "In Transit" example, reserving the strict graders for genuinely fixed outputs.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 11 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:187]: Made explicit in "Stack graders selectively" — stack only where path and answer can fail independently; one well-chosen grader for simple lookups and guardrail refusals.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+Thresholds made sense in an older world, where you tested a trained model against a huge, fixed set of utterances and a pass rate stood in for the exhaustive checking you could never do by hand. That number was always a stand-in for the maker's real question: is this good enough for what I need it to do? AI evaluations let you skip the stand-in. You lead with the qualitative definition of *good*, and each test carries its own explanation of why it passed or failed. That explanation is what you act on, not the pass rate: over a small, curated set, a pass rate isn't statistically meaningful, and it was never meant to be.
+
+Three questions replace the score, and the rest of this post is built to answer them:
+
+- **Coverage.** Does the test set represent what matters, including the edge cases and the things that should fail?
+- **Explanation.** When a test fails, why? Was it a real bug, an acceptable miss, or the agent correctly refusing?
```
Comment (adrianatruji) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:56]: For knowledge-grounded agents, I believe there's a failure mode worth naming here: Was the wrong source retrieved? This is to consider a knowledge-based agent can give a wrong answer not because of reasoning but because the retrieval surfaced the wrong document, or the corpus structure/format caused the right info to rank poorly. Even if this belongs to RAG evaluation territory, for agents that have a pure knowledge reteival function, this matters because the fix the user will have to apply is different (restructure knowledge, not rewrite instructions).

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+
+### Generate a starter set, then expand it with the template
+
+Copilot Studio's [agent evaluation](https://learn.microsoft.com/en-us/microsoft-copilot-studio/analytics-agent-evaluation-intro) in the Evaluate tab generates a starter set straight from your agent definition. Treat it as a starting point. Its quality tracks how well your agent and tools are described, which is exactly why the scoping work above pays off here. Simply reviewing what it generates already sharpens your scoping and guardrailing.
+
+To expand a valuable use case, use the template provided in the Evaluate tab. Export the evaluation template, which carries the request-and-expected-response format, add a short use-case description, a few sample questions, and your value ranges, and hand it to an LLM you already have. M365 Copilot will do it. Ask for about 20 new rows for that use case. Twenty is a good size for a test set: enough to represent the range, small enough to read and curate. The [hands-on pass at the end](#bonus-a-hands-on-pass) includes a ready-to-paste version of that prompt.
```
Comment (adrianatruji) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:126]: Not sure if providing 20 as a golden number feels like it contradicts the scoping guidance above. How many rows in test set we will need probably depends on the scope mentioned above: a narrow single-tool lookup might need less, a broad knowledge retrieval will need more. I would reframe it saying something like 20 rows per use case is a practical starting point, but understand "use case" as a narrowed concept (e.g. in the knowledge retrieval agent may be one of the subjects the agent provides answers on), rather than a blanket recommendation. That keeps it consistent with the earlier point that scope decides everything.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+
+"Is it ready?" tangles two different problems, and each needs a different fix.
+
+**A design problem: scope.** Agents work best with a goal and a scope. Handing an agent a tool is the core of how agents work, but a tool is not a scope: a single system of record serves many roles at once, and the whole of it is too broad to be one agent's job. The same orders database backs customer support, operations, and finance; a well-scoped agent takes one such role and harnesses whatever tools it needs, while a larger agent can hold several roles, each scoped and bounded so it can be evaluated on its own terms.
+
+**An agreement problem: what "good" means.** Before you evaluate anything, the business and the makers have to agree on what the agent does, what it does not do, and what *good* means for each role it plays. That agreement is the step most projects skip, and it's why "it's not ready" is a verdict no one can act on: nobody wrote down what ready would look like. Whether you're building an internal agent for colleagues or a partner handing a more complex agent to a customer for testing, the rule is the same: don't release an agent whose scope you cannot fully describe.
```
Comment (adrianatruji) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:37]: I think this is the right place to reinforce the message that the definition of "good" grounds on the business KPIs that the agent is meant to move (whether that's time saved, error reduction, or whatever OKR the organization is targeting with "agentifying" a process). I think it is understood between the lines, but we can be more explicit about the fact those business KPIs/OKRs are what ultimately tell the builders which use cases are must-do vs. nice-to-have.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+{: .prompt-tip }
+
+Then rerun the **same** set and use [**Compare with**](https://learn.microsoft.com/en-us/microsoft-copilot-studio/analytics-agent-evaluation-results#compare-test-results) to line the two runs up row by row. Watch the rows that moved the way you intended, and any that regressed because a change had a side effect, like the agent now guardrailing valid requests.
+
+A fix you can generalize is worth far more than one that clears a single row. Confirm it generalizes: add a few more variations of the same problem and check the agent handles them consistently. Run adjacent tests to check for unintended effects.
+
+### Iterate, then slot the rest for the next wave
+
+Iteration has a natural stopping point. Once the must-do core and the guardrails hold, that's the signal to stop, take stock of what's left, and slot it into the next version. Ship a defined V1, then grow the solution. It's an agile loop, and the test sets draw the line: everything green in scope ships, and everything still open becomes the backlog for the next wave. Once its shape settles, you can even [automate the loop from edit to test]({% post_url 2026-03-29-agentic-improvement-loop %}).
```
Comment (adrianatruji) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:202]: I think we need to acknowledge what happens when real users start consuming VS the test space we have been using so far. When users start consuming the agent, cases appear that may deserve the must-have status edge cases that couldn't be anticipated from test data, either because users weren't told clearly what the agent can/cannot do, so they judge it against wrong expectations; or there are field gaps (e.g. the business stakeholders' view didn't match the reality of how people actually work, so something that seemed like a nice-to-have turns out to be an adoption blocker). I think the whole process of what happens after this distinction may actually need its own section/space, but maybe we can mention high level!

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+{: .prompt-tip }
+
+Both problems share a root: you can't judge readiness without a scope and a definition of *good*. Write them down, then keep them in view by evaluating as you build. Start as early as the agent can answer a basic question; you don't need production data, a generated set from day one is enough to catch problems while they're cheap to fix. Then evaluate as the solution takes shape, each time you add a feature, cover another use case, or change something that shifts behavior like the model. Evaluation isn't a gate at the finish line, it's a signal you read all the way through, so the question shifts from "is my agent good?" to "is this version better than the last, and where?" Skip it and readiness stays a matter of opinion until delivery, the most expensive place to find out you and the business disagree. You're not adding work, you're moving it earlier, catching a weak spot while it's cheap instead of meeting it as poor behavior once users are already relying on it.
```
Comment (adrianatruji) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:42]: I think is right to say we don't need production data, but also acknowledge that well-designed test data should account for the complexity of production data as we get closer to the production line (messy formatting, edge values, missing fields, multi-language, etc), or otherwise the eval will overstate readiness. In similar lines to my other comment about what happens when real usage starts, I think this whole lifecycle in test VS prod environments may need broader space, but we can still call out this distinction

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 12 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:56]: For the purpose of this article, if the wrong source is retrieved, we  can bucket that under agent bug.  We have not enumerated types of agent bugs, there are many. And for knowledge in particular, there is no Grader that is adequate to evaluate source accuracy at the moment. There are some hacks that we can document, but this is a recurring ask that I expect will surface in-product very soon.  We can write a separate blog about that, it will be very very valuable. I have some perspectives already.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 13 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:126]: This is also a Product "limitation" where the evals are designed for a golden number of 20 with a max of 100.  so instead of making the test sets larger, they could scope use cases and goals  to keep the test size at ~20. **What if I just remove this that sounds inaccurate: "enough to represent the range,**". Now I'm just saying that 20 is a manageable size of testset.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 14 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:37]: The OKR is related to the selection of actual use case itself.  Ex: including this use case will save us  this much when it is used.   

For evaluating a use case, good  means:  "do I consider the use case covered (and the  OKR achieved), yes/no?" if yes, it is ready, if no, tweak the design to either cover it or to exclude it from this version of the agent.  
-> Is this unclear?  should I put this sentence in? I was really trying to stay away from Numbers and OKR talk.  

Note:  whether the goal is achieved is a subjective decision based on looking at the results and the  grader comments.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 15 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:202]: Decided offline: We'll write another article on the whole deployment process. 
Note: Today, that process is painful and does surface a lot of edge cases, but that's because most people don't have a good  eval based design strategy.  Hopefully, if this is followed, the next step will have fewer surprises.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 16 above)
Comment (KarimaKT) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:42]: Designing a really good grading strategy and scoping tests for a list of  common edge cases  (like multilanguage) can be a great  extra blog that goes into better worked out evaluation design examples. 

Best we can do in this article is look at the  test data generation prompt  section and include  some "messiness"  directives  that hints at th\is message.

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 13 above)
Comment (adrianatruji) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md:126]: Makes sense, Karima! Let me reflect it, thank you

Comment (adrianatruji) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]:
```suggestion
To expand a valuable use case, use the template provided in the Evaluate tab. Export the evaluation template, which carries the request-and-expected-response format, add a short use-case description, a few sample questions, and your value ranges, and hand it to an LLM you already have. M365 Copilot will do it. Ask for about 20 new rows for that use case. Twenty is a good size for a test set. The [hands-on pass at the end](#bonus-a-hands-on-pass) includes a ready-to-paste version of that prompt.
```

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 12 above)
Comment (adrianatruji) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: Like the idea, Karima! Agree with you it would be an in-product valuable element, keen to hear your thoughts on this one until then. Resolving!

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 15 above)
Comment (adrianatruji) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: Indeed! Thank you

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 14 above)
Comment (adrianatruji) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: Agree with you, Karima! I was thinking more outside the built-in evals in cases where, for example, the boundary of "what is good (enough)" comes for the business may be time-to-response (so maybe you would expect an answer from the agent that does not require multiple turns with the user trying to get to the outcome). But it's true this opens a door for other missing product capabilities like latency as a metric in the evals, and may not be desirable to open that door here in the context of the MCS article - so closing this one, and thank you!

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md (hunk 16 above)
Comment (adrianatruji) [reply _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]: I'll add a separate note on a line I think can account for this, in case you think it makes sense to include it, Karima. Thanks a lot!

File: _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+
+The graders that need an expected answer also need authoring effort, so spend it where it counts. Write out expected answers for your high-value use cases, run, adjust the design with generalizable fixes, and rerun. If a round of fixes touched a lot, regenerate similar tests to confirm the fixes held and generalized to the range you meant to cover.
+
+On a larger project, where a separate builder team delivers for the business SMEs, that authoring effort is shared work, not a solo chore. Agree up front on who does what: the business owners set the priority use cases and define what a good answer is, and the people who own the data supply the expected responses. From there, the makers expand those sets with AI to anticipate logical edge cases, build the solution iteratively against them, and deliver it with its expectations already described by the evaluations, so end-user testing holds few surprises.
```
Comment (adrianatruji) [inline _posts/2026-06-01-evaluation-driven-agent-readiness-copilot-studio.md]:
```suggestion
On a larger project, where a separate builder team delivers for the business SMEs, that authoring effort is shared work, not a solo chore. Agree up front on who does what: the business owners set the priority use cases and define what a good answer is, and the people who own the data supply the expected responses and flag the data variability (format inconsistencies, typical missing fields, etc.) that a good test strategy should account for. From there, the makers expand those sets with AI to anticipate logical edge cases, build the solution iteratively against them, and deliver it with its expectations already described by the evaluations, so end-user testing holds few surprises.
```

**Comment by KarimaKT** [comment]
@adilei, the final review feedback is now addressed in `7969fbf`. The post is complete, remains dated June 1 as intended, and is marked `published: true`. Once the refreshed preview check passes, could you please publish it? Thank you!
